// rest-config/config.js
const singleton = Symbol();
const singletonEnforcer = Symbol();
const path = require('path');
const yaml = require('js-yaml');
const fs   = require('fs');
const GLOBAL   = "GLOBAL";
const METADATA = "METADATA";
class Config {

    constructor(enforcer) {
        if (enforcer !== singletonEnforcer) {
            throw new Error('Cannot construct singleton');
        }
        this.data = null
    }

    static get instance() {
        if (!this[singleton]) {
            this[singleton] = new Config(singletonEnforcer);
        }

        return this[singleton];
    }
    static get rootPath() {
        return path.resolve(__dirname).replace('node_modules/rest-config', '')
    }
    load(files) {
        if (this.data != null) return;
        this.loadConfig(files[0], GLOBAL);
        this.loadConfig(files[1], METADATA);
    }
    loadConfig(path, name) {
        try {
            if (this.data == null) this.data = {}
            const doc = yaml.safeLoad(fs.readFileSync(path, 'utf8'));
            this.data[name] = doc;
        } catch (e) {
            throw e;
        }
    }
    loadGlobal(path) {
        this.loadConfig(path, GLOBAL);
    }
    loadMetadata(path) {
        this.loadConfig(path, METADATA);
    }

    get(name, key) {
        return this.data[name][key];
    }

    global(key) {
        return this.data[GLOBAL][key];
    }
    metadata(key) {
        return this.data[METADATA][key];
    }
}
module.exports = Config;
