// rest-core/server.js
const express = require('express');
const config = require('rest-config').instance;
const bodyParser = require('body-parser')
const GLOBAL_PATH = './global.yml';
const METADATA_PATH = './meta.yml';
const WebRouter = require('./src/router/web');
const ApiRouter = require('./src/router/api');
const ExtRouter = require('./src/router/ext');
const SwaggerClient = require('rest-swagger');
const Hook = require('./src/hook');
class Server {

    constructor(_rootPath) {
        config.load([GLOBAL_PATH, METADATA_PATH]);
        this.app = express();
        this.rootPath = _rootPath;
        new Hook();
    }
    setupMiddleware() {
        this.app.use(bodyParser.urlencoded({extended: false}));
        this.app.use(bodyParser.json());
    }
    setupRouter() {
        this.app.use('/', new WebRouter().getRouter());
        this.app.use('/doc', SwaggerClient.serve, SwaggerClient.setup());
        this.app.use('/api/ext', new ExtRouter(this.rootPath).getRouter()); 
        this.app.use('/api', new ApiRouter().getRouter()); 
    }
    listen() {
        this.app.listen(config.global('server')['port'], () => {
            console.log('Server is running at ... ' + config.global('server')['port']);
        });
    }
    start() {
        this.setupMiddleware();
        this.setupRouter();
        this.listen();
    }
}
module.exports = Server;
