// rest-core/src/controller/index.js
const Model = require('../model');

class Controller {
    constructor(resource) {
        this.model = new Model(resource);        
    }

    async getAll(req, res, next) {
        const data = await this.model.getAll();
        if (next) {
            res.data = data;
            next(req, res);
        } else {
            (typeof data !== "undefined") ?
            res.json({ message: 'success', data: data}) :
            res.json({ message: 'success', data: []});
        }
    }
    async get(req, res, next) {
        const data = await this.model.get(req.params.id);
        (typeof data !== "undefined") ?
            res.json({ message: 'success', data: data}) :
            res.json({ message: 'success', data: []});
    }
    async add(req, res, next) {
        const data = await this.model.add(req.body);
        (typeof data !== "undefined") ?
            res.json({ message: 'success', data: data}) :
            res.json({ message: 'success', data: []});
    }
    async update(req, res, next) {
        const data = await this.model.update(req.body);
        (typeof data !== "undefined") ?
            res.json({ message: 'success', data: data}) :
            res.json({ message: 'success', data: []});
    }
    async delete(req, res, next) {
        const data = await this.model.del(req.params.id);
        (typeof data !== "undefined") ?
            res.json({ message: 'success', data: data}) :
            res.json({ message: 'success', data: []});
    }

}
module.exports = Controller;
